File: iatrics_backend/src/simulation/consultationSimulator.js
const axios = require("axios");

class ConsultationSimulator {
  constructor(baseUrl) {
    this.baseUrl = baseUrl;
    this.userToken = null;
    this.providerToken = null;
    this.consultationId = null;
  }

  // =========================
  // LOGIN USER
  // =========================
  async loginUser() {
    const res = await axios.post(
      `${this.baseUrl}/api/auth/users/login`,
      {
        email: "user@test.com",
        password: "123456"
      }
    );

    this.userToken = res.data.data.token;
    return this.userToken;
  }

  // =========================
  // LOGIN PROVIDER
  // =========================
  async loginProvider() {
    const res = await axios.post(
      `${this.baseUrl}/api/auth/providers/login`,
      {
        email: "provider@test.com",
        password: "123456"
      }
    );

    this.providerToken = res.data.data.token;
    return this.providerToken;
  }

  // =========================
  // CREATE CONSULTATION
  // =========================
  async createConsultation() {
    const res = await axios.post(
      `${this.baseUrl}/api/consultations`,
      {
        providerId: 1,
        type: "video"
      },
      {
        headers: { Authorization: `Bearer ${this.userToken}` }
      }
    );

    this.consultationId = res.data.data.id;
    return this.consultationId;
  }

  // =========================
  // ACCEPT CONSULTATION
  // =========================
  async acceptConsultation() {
    return axios.post(
      `${this.baseUrl}/api/consultations/${this.consultationId}/accept`,
      {},
      {
        headers: { Authorization: `Bearer ${this.providerToken}` }
      }
    );
  }

  // =========================
  // GET AGORA TOKEN
  // =========================
  async getAgoraToken(role = "user") {
    const token = role === "user" ? this.userToken : this.providerToken;

    const res = await axios.get(
      `${this.baseUrl}/api/agora/token/${this.consultationId}`,
      {
        headers: { Authorization: `Bearer ${token}` }
      }
    );

    return res.data.data.token;
  }

  // =========================
  // END CONSULTATION
  // =========================
  async endConsultation() {
    return axios.post(
      `${this.baseUrl}/api/consultations/${this.consultationId}/end`,
      {},
      {
        headers: { Authorization: `Bearer ${this.providerToken}` }
      }
    );
  }

  // =========================
  // RUN FULL FLOW
  // =========================
  async run() {
    console.log("🚀 Starting Consultation Simulation...");

    await this.loginUser();
    console.log("✅ User logged in");

    await this.loginProvider();
    console.log("✅ Provider logged in");

    await this.createConsultation();
    console.log("📌 Consultation created:", this.consultationId);

    await this.acceptConsultation();
    console.log("🤝 Provider accepted");

    const userToken = await this.getAgoraToken("user");
    const providerToken = await this.getAgoraToken("provider");

    console.log("🎥 Agora tokens generated");

    // simulate call duration
    console.log("📞 Call started...");
    await new Promise((r) => setTimeout(r, 3000));

    await this.endConsultation();
    console.log("📴 Call ended");

    return {
      consultationId: this.consultationId,
      userToken,
      providerToken
    };
  }
}

module.exports = ConsultationSimulator;

File: iatrics_provider_app/lib/core/simulation/consultation_simulator.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/app_environment.dart';

class ConsultationSimulator {
  static final _controller = StreamController<String>.broadcast();
  static Stream<String> get stream => _controller.stream;

  static String get baseUrl => AppEnvironmentConfig.baseUrl;

  // =========================
  // 🚀 MAIN FLOW
  // =========================
  static Future<void> startFlow() async {
    try {
      print("🔐 USER LOGIN");

      final userLogin = await _post(
          "/api/auth/login", {"email": "user@test.com", "password": "123456"});

      final userToken = userLogin["token"];

      print("👨‍⚕️ PROVIDER LOGIN");

      final providerLogin = await _post("/api/auth/login",
          {"email": "provider@test.com", "password": "123456"});

      final providerToken = providerLogin["token"];

      print("📞 CREATE CONSULTATION");

      final createRes = await _post(
        "/api/consultations",
        {"providerId": "1", "type": "video"},
        token: userToken,
      );

      final consultationId = createRes["id"];

      _emit("CONSULTATION_CREATED");

      print("✅ PROVIDER ACCEPT");

      await _post(
        "/api/consultations/$consultationId/accept",
        {},
        token: providerToken,
      );

      _emit("IN_CALL");

      print("🎥 FETCH AGORA TOKENS");

      await _get("/api/agora/token/$consultationId", token: userToken);
      await _get("/api/agora/token/$consultationId", token: providerToken);

      print("🔚 END CALL");

      await _post(
        "/api/consultations/$consultationId/end",
        {},
        token: providerToken,
      );

      _emit("CALL_ENDED");

      print("✅ FLOW COMPLETED");
    } catch (e) {
      print("❌ SIMULATION ERROR: $e");
      _emit("ERROR");
      rethrow;
    }
  }

  // =========================
  // ⏳ WAIT HANDLER
  // =========================
  static Future<void> waitForCompletion({
    Duration timeout = const Duration(seconds: 15),
  }) async {
    await Future.delayed(timeout);
  }

  // =========================
  // 🌐 HTTP HELPERS
  // =========================
  static Future<Map<String, dynamic>> _post(
    String path,
    Map<String, dynamic> body, {
    String? token,
  }) async {
    final res = await http.post(
      Uri.parse("$baseUrl$path"),
      headers: {
        "Content-Type": "application/json",
        if (token != null) "Authorization": "Bearer $token",
      },
      body: jsonEncode(body),
    );

    _validate(res, path);

    return jsonDecode(res.body);
  }

  static Future<Map<String, dynamic>> _get(
    String path, {
    String? token,
  }) async {
    final res = await http.get(
      Uri.parse("$baseUrl$path"),
      headers: {
        if (token != null) "Authorization": "Bearer $token",
      },
    );

    _validate(res, path);

    return jsonDecode(res.body);
  }

  // =========================
  // 🔒 RESPONSE VALIDATION
  // =========================
  static void _validate(http.Response res, String path) {
    if (res.statusCode < 200 || res.statusCode >= 300) {
      throw Exception(
        "API FAILED [$path]: ${res.statusCode} → ${res.body}",
      );
    }
  }

  // =========================
  // 📡 EVENT EMITTER
  // =========================
  static void _emit(String event) {
    _controller.add(event);
  }
}
